### ISSUE TITLE
__attribute__((uncached)) in not applied to typedef'ed types

### ISSUE THREAD

**Issue by petrokarashchenko** [foss-for-synopsys-dwc-arc-processors/toolchain#221]
GCC Version: "GCC: (ARCompact/ARCv2 ISA elf32 toolchain 2019.09-rc3) 9.2.1 20191002"
Next code compiles with no warnings or errors.
```
typedef volatile struct {
    int a;
    char *b;
} __attribute__((uncached)) my_type_t;

void my_func(my_type_t *p)
{
    p->a = 10;
    p->b = NULL;
}

my_type_t var1;
int main (int argc, char ** argv)
{
    my_func(&var1);
    return 0;
}
```
Generated assembly code
```
   9               		.section	.text
  10               		.section	.text.my_func,"ax",@progbits
  11               		.align 4
  12               		.global	my_func
  14               	my_func:
  15 0000 D90A      		mov_s	r1,10	;0
  16 0002 A020      		st_s	r1,[r0]		;16
  17 0004 D900      		mov_s	r1,0	;0
  18 0006 A021      		st_s	r1,[r0,4]		;16
  19 0008 7EE0      		j_s	[blink]
```
Expected assembly code
```
   9               		.section	.text
  10               		.section	.text.my_func,"ax",@progbits
  11               		.align 4
  12               		.global	my_func
  14               	my_func:
  15 0000 D90A      		mov_s	r1,10	;0
  16 0002 A020      		st.di	r1,[r0]		;16
  17 0004 D900      		mov_s	r1,0	;0
  18 0006 A021      		st.di	r1,[r0,4]		;16
  19 0008 7EE0      		j_s	[blink]
```
It is expected that `st.di`/`ld.di` instructions are used for types with `__attribute__((uncached))` attribute

**Comment by petrokarashchenko** [comment]
Is it possible to fix this is in GNU Toolchain for ARC Processors, 2019.09 release?

**Comment by abrodkin** [comment]
@petrokarashchenko could you please elaborate a bit on what is a project you're working on where this toolchain versions is going to be used and if there're any deadline we need to worry about working on the fix?

**Comment by claziss** [comment]
Hi,

In your example, the pointer `my_type_t *p` is defined as `uncached` thus, the compiler does exactly what it should do. For example, considering the fallowing snip:

```c
my_type_t foo (my_type_t *p)
{
  return *p++;
}
```
you will get:
```asm
foo:
        ldd.di r2,[r1]
        j_s.d   [blink]
        std r2,[r0]
        .size   foo, .-foo
```
which it is what it is expected.

If you want the members of your structure to be uncached, then I would do this:

```c
typedef volatile struct {
    __attribute__((uncached)) int a;
    __attribute__((uncached)) char *b;
}  my_type_t;

void my_func(my_type_t *p)
{
    p->a = 10;
    p->b = 0;
}
```

with the result:
```asm
my_func:
        st.di   10,[r0] ;37
        st.di   0,[r0,4]        ;37
        j_s     [blink]
        .size   my_func, .-my_func
```

Probably the latter is what you were expecting :)

**Comment by petrokarashchenko** [comment]
@claziss in case of
```
typedef volatile struct {
    __attribute__((uncached)) int a;
    __attribute__((uncached)) char *b;
}  my_type_t;
```
I do not fully understand does `__attribute__((uncached)) char *b` means that `p->b` access is uncached or `*p->b` access is uncached.

I make analogy with GCC `__attribute__((packed))`.

Actually my assumptions are next:
```
typedef volatile struct {
    int a;
    char *b;
} __attribute__((uncached)) my_type_t;

typedef volatile struct {
    my_type_t a;
    my_type_t *b;
    int c;
} my_type1_t;

void my_func(my_type_t *p, my_type1_t *p1)
{
    p->a = 10; // access uncached
    p->b = NULL; // access uncached
    *p->b = 'a'; // access is cached

    p1->a.a = 10; // access is uncached
    p1->b = NULL; // access is cached
    p1->b.a = 10; // access is uncached
    p1->c = 10; // access is cached
}
```
@claziss looking forward for your input

**Comment by petrokarashchenko** [comment]
@claziss you wrote that: "In your example, the pointer `my_type_t *p` is defined as `uncached`", but
```
typedef volatile struct {
    int a;
    char *b;
} __attribute__((uncached)) my_type_t;
```
Does not state anything about pointers.

If I use
```
my_type_t a;

void my_func(void)
{
    a.a = 10;
    a.b = 0;
}
```
I still expect that `st.di`/`ld.di` instructions are used.

**Comment by petrokarashchenko** [comment]
@abrodkin I'm working on project that use 4xCore ARC600. We want to use any working GNU toolchain. I have spend quite a lot of time trying to bring-up code with `GNU Toolchain for ARC Processors, 2019.03`, but `GNU Toolchain for ARC Processors, 2019.03` has a bug with usage of `-mmul64` option (wrong assembly code is generated for some `mul64` cases) and usage of `__attribute__((interrupt ("ilink1")))` (that is fixed in current RC).
The deadlines we have is end of April 2020 to start beta testing (and that assumes that we pass all internal testing and development is stable).
Let me describe a use case a bit more:
I have 4xCore CPU and want to use share memory IPC, so I expect to allocate a region of memory and memory map it into `uncached` typedef'ed struct. In this case a writing end will write structure fields with `st.di` instructions and reading end will read structure fields with `ld.di` instructions, so I can bypass flush/invalidate cache.

**Comment by claziss** [comment]
I see, u expect `uncached` type attribute to be propagated to each member of the struct, union, or class, which doesn't happen. Ok, this is a "feature" :)

**Comment by petrokarashchenko** [comment]
@claziss I do not agree.
I see the test case
```
/* Check 'uncached' type attribute.  */

typedef volatile unsigned int RwReg  __attribute__ ((uncached));

typedef struct {
  RwReg UART_THR;
  int SIDE_DISH;
} UART;

void uart_putc(UART *port, char c)
{
    port->UART_THR = c;
    port->SIDE_DISH = c;
}
```
In the example `__attribute__ ((uncached))` is applied to `RwReg` type. @claziss do you agree?
If `__attribute__ ((uncached))` is applied to `RwReg` type, then it is expected that
```
RwReg a;
RwReg *b;

a = 10;
*b = 10;
```
will generate `uncached` access. @claziss agree? That means that all memory region occupied by `RwReg` type (aka memory block of `sizeof(RwReg)`) is accessed using `st.di`/`ld.di` instructions.

Going further I use the same for complex types by
```
typedef volatile struct {
    int a;
    char *b;
} __attribute__((uncached)) my_type_t;
```
And assume that all all memory region occupied by `my_type_t` type (aka memory block of `sizeof(my_type_t)`) is accessed using `st.di`/`ld.di` instructions as well.

So I assume reported issues is a "bug", not a "feature". If we didn't have `__attribute__((uncached))` at all then implementing it in proper way would be a "feature" :)

**Comment by abrodkin** [comment]
@petrokarashchenko I guess you're talking about this one https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/issues/73, right?

Ok given your deadline in April indeed if we ever need to **fix it** then it won't wait for the next release we're going to have sometime during spring 2020. But looks like there's a work-around proposed by @claziss so maybe it will be enough for you now.

Also speaking about your initial explanation and further follow-up by @claziss do you see that kind of behavior (attribute propagation to structure members) when compile for any other architectures?

**Comment by petrokarashchenko** [comment]
@abrodkin I have tried the workaround that @claziss suggest, but I have next use case that I expect to work:
```
typedef volatile struct {
    int a;
    long long b;
} __attribute__((uncached)) my_type_t;

typedef volatile struct {
    my_type_t *a1;
    my_type_t *a2 ;
} __attribute__((uncached)) my_type1_t;

void my_func(void)
{
    my_type1_t *p = (my_type1_t *)alloc_shared(sizeof(my_type1_t));
    p->a1 = (my_type_t *)alloc_shared(sizeof(my_type_t));
    p->a2 = (my_type_t *)alloc_shared(sizeof(my_type_t));

    p->a1->a = 10;
    p->a1->b = 20;
    p->a2->a = 30;
    p->a2->b = 40;

    // send p valued over mailbox
    cpu_mailbox_send(p);
}
```
So I expect that I can access `p->a1` value (address) in uncached manner and `p->a1->a` in uncached manner.

When I use
```
typedef volatile struct {
    __attribute__((uncached)) int a;
    __attribute__((uncached)) long long b;
}  my_type_t;

typedef volatile struct {
    __attribute__((uncached)) my_type_t *a1;
    __attribute__((uncached)) my_type_t *a2 ;
} my_type1_t;
```
I do not know is `p->a1` access is uncached or not. Or maybe I need to use
```
typedef volatile struct {
    __attribute__((uncached)) int a;
    __attribute__((uncached)) long long b;
} __attribute__((uncached)) my_type_t;

typedef volatile struct {
    __attribute__((uncached)) my_type_t *a1;
    __attribute__((uncached)) my_type_t *a2 ;
} __attribute__((uncached)) my_type1_t;
```
What comes from
```
typdef enum {
    e1,
    e2
} e_t;

typedef volatile struct {
    __attribute__((uncached)) e_t a;
    __attribute__((uncached)) long long b;
} __attribute__((uncached)) my_type_t;
```
Actually I get message that attribute for enum is ignored...
Does this mean that I need to
```
typdef enum {
    e1,
    e2
} __attribute__((uncached)) e_t;

typedef volatile struct {
     e_t a;
    __attribute__((uncached)) long long b;
} __attribute__((uncached)) my_type_t;
```
??? If yes then why I need to add `__attribute__((uncached))` to enum `typedef ` if this is a value only and I use same type in other places for cached access.

I mean that I'm just not sure what code I will get at the end.

**Comment by claziss** [comment]
Please let me know if this patch fixes this issue:
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/e69526effeb578e0fcda4d849bc0875b31558be2

**Comment by petrokarashchenko** [comment]
@claziss according to test case I expect that specified patch should fix the issue

**Comment by petrokarashchenko** [comment]
I will wait for next toolchain build to verify

**Comment by petrokarashchenko** [comment]
@claziss is it possible to get pre-built "debug" version of ARC GCC toolchain containing the fix? So I can fully verify the fix. I need baremetal big-endian variant for Linux x86_64 or MAC.

**Comment by claziss** [comment]
You can build it yourself :) for both MacOS or Linux. For win is a bit more complex :). The howto is in the READ.me of this project. Please let us know your experience, and what will you change.
P.S. I may do a brew formula for MacOS but not now...

**Comment by petrokarashchenko** [comment]
@claziss I have made some experiments and fix works for simple types like enums.
But case:
```
typedef enum {
    e1,
    e2
} my_enum_t;

typedef struct {
    int a;
    int *p;
} my_struct_t;

typedef volatile struct {
	my_enum_t a;
	my_struct_t b;
} __attribute__((uncached)) my_type_t;

void my_func(my_type_t *p)
{
    p->a = e2;
    p->b.a = 10;
    p->b.p = NULL;
    *p->b.p = 10;
}
```
Generates
```
  33 0016 78E0      		.section	.text.my_func,"ax",@progbits
  34               		.align 4
  35               		.global	my_func
  37               	my_func:
  38 0000 D901      		mov_s	r1,1	;0
  39 0002 1800 0060 		st.di	r1,[r0]	;26
  40 0006 D90A      		mov_s	r1,10	;0
  41 0008 A021      		st_s	r1,[r0,4]		;16
  42 000a DA00      		mov_s	r2,0	;0
  43 000c A042      		st_s	r2,[r0,8]		;16
  44 000e 8002      		ld_s	r0,[r0,8]		;15
  45 0010 A020      		st_s	r1,[r0]		;16
  46 0012 7EE0      		j_s	[blink]
```
So `uncached` is not propagated to all region of memory defined by `my_type_t `.
Could you please take a look is it doable to fix this as well?

**Comment by petrokarashchenko** [comment]
If I make
```
typedef enum {
    e1,
    e2
} my_enum_t;

typedef struct {
    int a;
    int *p;
} my_struct_t;

typedef volatile struct {
    my_enum_t a;
    my_struct_t b;
} __attribute__((uncached)) my_type_t;
```
I get `warning: ignoring attributes applied to 'struct <anonymous>' after definition`
O understand that it is possible to workaround using
```
typedef enum {
    e1,
    e2
} my_enum_t;

typedef struct {
    int a;
    int *p;
} __attribute__((uncached)) my_struct_t;

typedef volatile struct {
    my_enum_t a;
    my_struct_t b;
} __attribute__((uncached)) my_type_t;
```
but `my_struct_t` by itself (like my_enum_t) is not supposed to be `uncached`. Only block of memory accessed by `my_type_t` is supposed to be `uncached`

**Comment by petrokarashchenko** [comment]
First of all I want to thank you @claziss and @abrodkin for taking this effort and implement the change in this release. Thank you very much for support and hope that it will be possible to implement behaviour described above with minimal effort. If not then I will need to think about possible workarounds.

**Comment by petrokarashchenko** [comment]
@claziss I have applied next change and seems to get the behaviour described above.
```
diff --git a/gcc/config/arc/arc.c b/gcc/config/arc/arc.c
index c44eaae9346..f2d2f6743ed 100644
--- a/gcc/config/arc/arc.c
+++ b/gcc/config/arc/arc.c
@@ -11268,23 +11268,30 @@ arc_is_uncached_mem_p (rtx pat)
     {
       attrs = TYPE_ATTRIBUTES (TREE_TYPE (addr));
       if (lookup_attribute ("uncached", attrs))
-	return true;
-
-      attrs = TYPE_ATTRIBUTES (TREE_TYPE (TREE_OPERAND (addr, 0)));
-      if (lookup_attribute ("uncached", attrs))
-	return true;
-    }
+       return true;

-  if (TREE_CODE (addr) == COMPONENT_REF)
-    {
-      /* First, use the FIELD_DECL from tree operand 1.  */
-      attrs = TYPE_ATTRIBUTES (TREE_TYPE (TREE_OPERAND (addr, 1)));
-      if (lookup_attribute ("uncached", attrs))
-	return true;
-      /* Second, use the FIELD_DECL from tree operand 0.  */
       attrs = TYPE_ATTRIBUTES (TREE_TYPE (TREE_OPERAND (addr, 0)));
       if (lookup_attribute ("uncached", attrs))
-	return true;
+        return true;
+    }
+
+  if (TREE_CODE (addr) == COMPONENT_REF
+      || TREE_CODE (addr) == ARRAY_REF
+      || TREE_CODE (addr) == ARRAY_RANGE_REF)
+    {
+      while (handled_component_p (addr))
+        {
+          /* First, use the FIELD_DECL from tree operand 1.  */
+          attrs = TYPE_ATTRIBUTES (TREE_TYPE (TREE_OPERAND (addr, 1)));
+          if (lookup_attribute ("uncached", attrs))
+            return true;
+          /* Second, use the FIELD_DECL from tree operand 0.  */
+          attrs = TYPE_ATTRIBUTES (TREE_TYPE (TREE_OPERAND (addr, 0)));
+          if (lookup_attribute ("uncached", attrs))
+            return true;
+
+          addr = TREE_OPERAND (addr, 0);
+        }
     }
   return false;
 }
```
Please take a look and say what do you think.
We can add test case from example above to verify. Of even extend it with next 
```
typedef enum {
    e1,
    e2
} my_enum_t;

typedef struct {
    int a;
    int *p;
} my_struct_t;

typedef volatile struct {
    my_enum_t a;
    my_struct_t b;
} __attribute__((uncached)) my_type_t;

typedef volatile struct {
    my_enum_t a;
    my_struct_t b;
} my_type1_t;

void my_func(my_type_t *p, my_type1_t *p1)
{
    p->a = e2;
    p->b.a = 10;
    p->b.p = NULL;
    *p->b.p = 10;

    p1->a = e2;
    p1->b.a = 10;
    p1->b.p = NULL;
    *p1->b.p = 10;
}
```
and verify code
```
  10               		.section	.text.my_func,"ax",@progbits
  11               		.align 4
  12               		.global	my_func
  14               	my_func:
  15 0000 DC01      		mov_s	r12,1	;0
  16 0002 1800 0320 		st.di	r12,[r0]	;26
  17 0006 DA0A      		mov_s	r2,10	;0
  18 0008 1804 00A0 		st.di	r2,[r0,4]	;26
  19 000c DB00      		mov_s	r3,0	;0
  20 000e 1808 00E0 		st.di	r3,[r0,8]	;26
  21 0012 1008 0800 		ld.di	r0,[r0,8]	;23
  22 0016 A040      		st_s	r2,[r0]		;16
  23 0018 A180      		st_s	r12,[r1]		;16
  24 001a A141      		st_s	r2,[r1,4]		;16
  25 001c A162      		st_s	r3,[r1,8]		;16
  26 001e 8102      		ld_s	r0,[r1,8]		;15
  27 0020 A040      		st_s	r2,[r0]		;16
  28 0022 7EE0      		j_s	[blink]
```

**Comment by petrokarashchenko** [comment]
@claziss Is it possible to include above fix into 2019.09 release?

**Comment by abrodkin** [comment]
@petrokarashchenko please find new toolchain binaries published as `arc-2019.09-rc4` here https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/releases/tag/arc-2019.09-rc4 and let us know if that's sufficient for your project or not.

Given what we discuss here is not strictly a bug but more of an enhancement if not a new feature I don't want to gate this toolchain release by that long discussions, implementation and re-verification. Especially since we're talking about GNU toolchain which you may rebuild on your own with whatever  modifications on top of the any release.

Please let me know what do you think.

**Comment by petrokarashchenko** [comment]
@abrodkin That makes sense. I fully understand that you have a long re-verification process for all the changes and that 2019.09 milestone is overdue. You are 100% right that if we are talking about GNU toolchain I can do any modifications, but I'm not a compiler expert, and do not have any infrastructure to verify changes that I propose (in terms of full regression), so that is the biggest pain for me.

I think that current enchantment is a big step forward and thank you for integrating it into 2019.09. Hope to continue discussion and include proposed changes into next release gate. Maybe continue discussion in a separate issue.

I think I will take `arc-2019.09-rc4` and change my code by adding some ugly additional typedefs to handle situation that I have described and hope to work together with toolchain maintainers to incorporate full solution for enchantment. If I will get a green light I will take a `arc-2019.09` final source code and add a patch on top of that to build toolchain that fits my needs in the best way.

**Comment by petrokarashchenko** [comment]
In general I need only some help just to be sure that we are on the same page.
I used next steps to build toolchain locally:
```
1. clone all repos according to description in README.md
2. cd toolchain
3. git checkout arc-dev
4. ./arc-versions.sh
5. ./build-all.sh --no-uclibc --install-dir /workspace/install --no-multilib --cpu arc600 --target-cflags "-Os -g" --big-endian
```
but at the end I get not a tarball like pre-build toolchain, but a directory `/workspace/install` with folder structure.
I only would like to get some help with commands you guys use to get `arc_gnu_2019.09-rc3_prebuilt_elf32_be_linux_install.tar.gz` for example, so I can do the same and get the same image.

**Comment by petrokarashchenko** [comment]
I have created PR https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/pull/82 for better tracking

**Comment by abrodkin** [comment]
> I only would like to get some help with commands you guys use to get `arc_gnu_2019.09-rc3_prebuilt_elf32_be_linux_install.tar.gz` for example, so I can do the same and get the same image.

We just tarball entire folder (`cd arc-elf32-xxx && tar czf arc-elf32-xxx.tar.gz arc-elf32-xxx/`), you may see it here: https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/blob/arc-releases/release.mk#L177.

**Comment by abrodkin** [comment]
@claziss, @petrokarashchenko I guess we need to bump a milestone to `2020.03` as proper fix was not done yet and then it's not in 2019.09, right?

**Comment by petrokarashchenko** [comment]
I have closed PR that I have made. Changes were dummy ;)
It handled struct members, but case
```
typedef volatile __attribute__((uncached)) int uc_arr_int[20];
typedef volatile __attribute__((uncached)) int uc_int;

volatile __attribute__((uncached)) int s[20];
volatile __attribute__((uncached)) int b;

uc_arr_int su;
uc_int bu;

void foo(void)
{
    s[10] = 15;
    b = 10;

    su[10] = 15;
    bu = 10;
}
```
still generates wrong code
```
  40               		.global	foo
  42               	foo:
  43 0000 1CFC B6C8 		st.a	fp,[sp,-4]	;26
  44 0004 230A 3700 		mov	fp,sp	;5
  45 0008 224A 03C0 		mov	r2,15	;5
  46 000c 1E00 7080 		st	r2,[@s+40]	;26
  46      0000 0000 
  47 0014 224A 0280 		mov	r2,10	;5
  48 0018 1E00 7080 		st	r2,[@b]	;26
  48      0000 0000 
  49 0020 224A 03C0 		mov	r2,15	;5
  50 0024 1E00 7080 		st	r2,[@su+40]	;26
  50      0000 0000 
  51 002c 224A 0280 		mov	r2,10	;5
  52 0030 1E00 7080 		st	r2,[@bu]	;26
  52      0000 0000 
  53 0038 78E0      		nop_s
  54 003a 1404 341B 		ld.ab	fp,[sp,4]	;23
  55 003e 7EE0      		j_s	[blink]
```

Looking forward to get proper fix from @claziss

**Comment by claziss** [comment]
This patch should resolve this:
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/2080016aa654d7d4b8dc4fddc1e151d5ffa35d54

**Comment by petrokarashchenko** [comment]
@claziss thank you very much for such huge improvement.
I have tested most of the code parts that had problems. Those are fine.

But now I'm getting some strange issue. I have next code in `test.c`
```
#include <stdio.h>

#define MY_ASSERT_OPT( condition, opt )                        \
    if (!(condition))                                           \
    {                                                           \
        printf("MY_ASSERT {%s:%d} opt=%d\n", __FILE__, __LINE__, opt);   \
        f0 (1, 2, __LINE__);               \
        while(1);                                               \
    }

typedef struct {
    uint8_t a;
    uint8_t b;
} my_type;

static my_type * f2 (int p1, int p2);

void f0(int p1, int p2, int p3)
{

}

void f1(int p1, int p2, int p3)
{
    my_type *c;

    c = f2 (p1, p2);

    MY_ASSERT_OPT (c->a == 1, (p1 << 4) | p2);
}

static my_type ref[4];

static my_type * f2 (int p1, int p2)
{
    MY_ASSERT_OPT(p1 == 1, (p1 << 4) | p2);
    return (&ref[p2]);
}
```
The compilation of `test.c` never ends.
My compiler options are next:
```
CFLAGS   = -mcpu=arc600 -mtune=arc600 -mbig-endian -mmul64 -mbarrel-shifter
CFLAGS  += -ffunction-sections -fdata-sections -fno-common -fno-strict-aliasing -c
CFLAGS  += -Os -Wa,-adhlns="$@.lst"
```

**Comment by petrokarashchenko** [comment]
Changing `-Os` to `-O2` makes issue disappear.

**Comment by petrokarashchenko** [comment]
But with `O2` I'm getting compilation stuck in other files. Need to evaluate code samples

**Comment by petrokarashchenko** [comment]
All my issues related to `uncached` attribute are fixed using latest commits from `2020.03` gcc branch

**Comment by claziss** [comment]
@petrokarashchenko I've rebased/merged/fixed the 2030.03. Please refer to the latest. I'll close this issue, and keep only one.

**Comment by abrodkin** [comment]
Fixed by https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/fc5aa9c7be10d21206537161f2245c097e06f456 in `arc-2020.03`.
